lib/grandpa/message_handler.go

```diff
@@ -299,20 +299,43 @@ func getEquivocatoryVoters(votes []AuthData) map[ed25519.PublicKeyBytes]struct{}
 	return eqvVoters
 }
 
+func verifyIfDescendantOfHighestFinalisedBlock(blockState BlockState, hash common.Hash) error {
+	highestHeader, err := blockState.GetHighestFinalisedHeader()
+	if err != nil {
+		return fmt.Errorf("could not get highest finalised header: %w", err)
+	}
+
+	isDescendant, err := blockState.IsDescendantOf(highestHeader.Hash(), hash)
+	if err != nil {
+		return fmt.Errorf("could not check if %s is descendant of %s: %w", hash, highestHeader.Hash(), err)
+	}
+	if !isDescendant {
+		// rename this error
+		return errInvalidVoteBlock
+	}
+
+	return nil
+
+}
+
 func (h *MessageHandler) verifyCommitMessageJustification(fm *CommitMessage) error {
 	if len(fm.Precommits) != len(fm.AuthData) {
 		return ErrPrecommitSignatureMismatch
 	}
 
+	if fm.SetID != h.grandpa.state.setID {
```

Comment:
A bit out of scope, but why is it named `fm`??

Reply:
no clue!